docs/Plugins-Guide.md

```diff
@@ -0,0 +1,232 @@
+
+`decorate` is not the unique api that you can use to extend the server functionalities, you can also use `decorateRequest` and `decorateReply`, what they do, is pretty obvious.
+
+We've seen how extend server functionalities and how handle the encapsulation system, but what if you need to add a functions that myst be executed every time that the server "[emits](https://github.com/fastify/fastify/blob/master/docs/Lifecycle.md)" an event?
+
+<a name="hooks"></a>
+## Hooks
+You just built and amazing utility, but now you need to execute that for every request, this is what you will likely do:
+```js
+fastify.decorate('util', (req, key, value) => { req.key = value })
+
+fastify.get('/plugin1', (req, reply) => {
+  fastify.util(req, 'timestamp', new Date())
+  reply.send(req)
+})
+
+fastify.get('/plugin2', (req, reply) => {
+  fastify.util(req, 'timestamp', new Date())
+  reply.send(req)
+})
+```
+I think we all agree that this is terrible. Code repeat, awful readability and it cannot scale.
+
+So what can you do to avoid this annoying issue? Yes, you are right, use an [hook](https://github.com/fastify/fastify/blob/master/docs/Hooks.md)!  
+```js
+fastify.decorate('util', (req, key, value) => { req.key = value })
+
+fastify.addHook('preHandler', (req, reply, done) => {
+  fastify.util(req, 'timestamp', new Date())
+  done()
+})
+
+fastify.get('/plugin1', (req, reply) => {
+  reply.send(req)
+})
+
+fastify.get('/plugin2', (req, reply) => {
+  reply.send(req)
+})
+```
+Now for every request you will run your utility, it is obvious that you can register as many hooks as you need.  
+It can happen that you want that a hook must be executed just for a subset of routes, how can you do that?  Yep, encapsulation!
+
+```js
+fastify.register((instance, opts, next) => {
+  instance.decorate('util', (req, key, value) => { req.key = value })
+
+  instance.addHook('preHandler', (req, reply, done) => {
+    instance.util(req, 'timestamp', new Date())
+    done()
+  })
+
+  instance.get('/plugin1', (req, reply) => {
+    reply.send(req)
+  })
+
+  next()
+})
+
+fastify.get('/plugin2', (req, reply) => {
+  reply.send(req)
+})
+```
+Now your hook will run just for the first route!
+
+As you probably noticed at this time, `request` and `reply` are not the standard Nodejs *request* and *response* objects, but Fastify's objects.  
+Let's say that you are arriving from a framework like Express or Restify, and you already have some Middleware that do exactly what you need, and you don't want to redo al the work.
+
+<a name="middlewares"></a>
+## Middlewares
+Fastify [supports](https://github.com/fastify/fastify/blob/master/docs/Middlewares.md) out of the box Express/Restify middlewares this means that you can just drop-in your old code and it will work! (faster, by the way)  
+How we can do that? Checkout our middlewares engine, [middie](https://github.com/fastify/middie).
+```js
+const yourMiddleware = require('your-middleware')
+fastify.use(yourMiddleware)
+```
```

Comment:
We need to note that middlewares do not respect encapsulation, and they are executed before routing.